### PR TITLE
adding keep in memory

### PR DESCRIPTION
**Pull request by Mustapha-AJEGHRIR**
Fixing #514 .

Hello @mariosasko 👋, I have implemented what you have recommanded to fix the keep in memory problem for shuffle on the issue #514 .

### CODE REVIEW

File: src/datasets/arrow_dataset.py [hunk 1]
```diff
@@ -3585,7 +3588,7 @@ def shuffle(
             generator = np.random.default_rng(seed)
 
         # Check if we've already cached this computation (indexed by a hash)
-        if self.cache_files:
+        if self.cache_files and not keep_in_memory:
```
Comment (mariosasko) [inline src/datasets/arrow_dataset.py:3591]: This line should stay the same:

```suggestion
        if self.cache_files:
```

Instead, [this](https://github.com/huggingface/datasets/blob/7d51531c58127d44565d6d1a6a7aa75e55d2b036/src/datasets/arrow_dataset.py#L3603) line should be replaced with:
```python
indices_cache_file_name=indices_cache_file_name if not keep_in_memory else None
```

**Review by mariosasko — COMMENTED** [review]
Thanks!

We also need a test with `shuffle(keep_in_memory=True)` for this fix. Let me know if you need help with that.

**Comment by Mustapha-AJEGHRIR** [comment]
Hi @mariosasko , I have added a test for the `keep_in_memory` version. I have also removed the `Compatible with temp_seed` part in the scope of `dset_shuffled`, please verify if that makes sense.

**Review by mariosasko — APPROVED** [review]
Looks all good now!

**Review by lhoestq — APPROVED** [review]
Nice !

**Pull request merged**